### PR TITLE
feat: add pulsar producer support to enable message publishing

### PR DESCRIPTION
**Pull request by fracasula**
## Description of the change

Adding Pulsar producer to write directly into Pulsar.

Useful for testing Pulsar capabilities without intermediaries.

### Additions
* made sources optionally as an integer for scenarios where we don't really need write keys (e.g. pushing straight to Pulsar)

### CODE REVIEW

File: cmd/producer/functions.go [hunk 1]
```diff
@@ -336,3 +372,56 @@ func optionalMap(s string, items []string) []int {
...
+func getHostname(hostname string) (string, int, error) {
+	match := hostnameRE.FindStringSubmatch(hostname)
+	if len(match) <= 2 {
+		return "", 0, fmt.Errorf("hostname is invalid: %s", hostname)
+	}
+
+	instanceNumber, err := strconv.Atoi(match[1])
+	if err != nil {
+		return "", 0, fmt.Errorf("error getting instance number from hostname %s: %v", hostname, err)
+	}
```
Comment (pChondros) [inline cmd/producer/functions.go:419]: Can we add some tests here as the hostname is an issue encountered in the past?

File: cmd/producer/functions.go [hunk 2]
```diff
@@ -336,3 +372,56 @@ func optionalMap(s string, items []string) []int {
...
+
+	deploymentName := match[2]
+	if deploymentName == "" {
+		return "", 0, fmt.Errorf("deployment name is empty: %s", hostname)
+	}
```
Comment (pChondros) [inline cmd/producer/functions.go:424]: Can we add some tests here as the hostname is an issue encountered in the past? Sorry for the duplicate :)

File: internal/producer/pulsar.go
Comment (pChondros) [inline internal/producer/pulsar.go:1]: Have you intentionally not tested the cases that codecov mentioned in this file?

**Review by pChondros — APPROVED** [review]

File: internal/producer/pulsar.go
Comment (fracasula) [reply internal/producer/pulsar.go:1]: Nope, I'll check them out thanks.

File: cmd/producer/main.go [hunk 3]
```diff
@@ -58,15 +69,15 @@ type publisherCloser interface {
 }
 
 func main() {
-	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt, os.Kill)
```
Comment (ASaiAnudeep) [inline cmd/producer/main.go]: Is there a specific reason for switching from `syscall` to `os`?
I had previously updated it to use `syscall` https://github.com/rudderlabs/rudder-load/commit/cf6d60b11cba7a83ba3fb95ae7a4380554aaf5c9 because we were facing issues in handling `SIGTERM`, which caused requests failing with a 499 status.

**Review by ASaiAnudeep — APPROVED** [review]

File: cmd/producer/main.go (hunk 3 above)
Comment (fracasula) [reply cmd/producer/main.go]: Compatibility depending on OS since we might run these script on the devs machines I thought I'd use the `os` package instead. I wasn't aware of the issue you came across to though. I'll add back `syscall.SIGTERM` as well then so effectively we'll support `SIGINT`, `SIGKILL` and `SIGTERM` :+1: 

Thanks for pointing this out :bow:

File: cmd/producer/functions.go (hunk 1 above)
Comment (fracasula) [reply cmd/producer/functions.go:419]: Is the `TestHostname` not good enough? https://github.com/rudderlabs/rudder-load/blob/pulsar-producer/cmd/producer/main_test.go#L782

Comment (pChondros) [reply cmd/producer/functions.go:419]: As we discussed this is guarded by the regex so let's skip the test for this.
